### PR TITLE
Check metrics details from `raw_connect`

### PR DESCRIPTION
**Pull request by jhernand**
Currently the `raw_connect` doesn't check the metrics connection
details. This patch adds that check.
    
This patch fixes partially the following bug:
 
  RHV provider metrics validation is broken
  https://bugzilla.redhat.com/1510533

### CODE REVIEW

**Comment by jhernand** [comment]
@pkliczewski @borod108 @masayag please review.

Take into account that this pull request should be merged after ManageIQ/manageiq-providers-ovirt#132. This currently contains two commits, the first is also in that other pull request. So please focus only in the second commit of this pull request.

**Comment by jhernand** [comment]
Note also that for this to work the UI has to be modified to send the required data. That is pull request ManageIQ/manageiq-ui-classic#2651.

**Comment by pkliczewski** [comment]
@jhernand This PR requires few changes in vcr files. For now the specs are failing.

**Comment by jhernand** [comment]
[Bug 1509432 - Failed validation when adding RHV provider](https://trello.com/c/KiNKHJbL/235-bug-1509432-failed-validation-when-adding-rhv-provider)

File: app/models/manageiq/providers/redhat/infra_manager/api_integration.rb [hunk 1]
```diff
@@ -306,6 +309,44 @@ def raw_connect(opts = {})
...
+        :port     => port,
+        :database => database
+      }
+      begin
+        OvirtMetrics.connect(opts)
+        OvirtMetrics.connected?
+      rescue StandardError => error
+        raise adapt_metrics_error(error)
+      ensure
+        OvirtMetrics.disconnect rescue nil
```
Comment (masayag) [inline app/models/manageiq/providers/redhat/infra_manager/api_integration.rb]: seems that rubocop would prefer 

        begin
          OvirtMetrics.disconnect
        rescue
          nil
        end

but we have the same style in line 145

**Comment by oourfali** [comment]
@miq-bot add_label gaprindashvili/yes

File: app/models/manageiq/providers/redhat/infra_manager/api_integration.rb (hunk 1 above)
Comment (jhernand) [reply app/models/manageiq/providers/redhat/infra_manager/api_integration.rb]: I will do as Rubocop says.

File: app/models/manageiq/providers/redhat/infra_manager/api_integration.rb [hunk 2]
```diff
@@ -232,17 +219,132 @@ def process_api_features_support
...
+    # @option opts [String] :ca_certs The custom trusted CA certificates used to check the TLS certificates of the
+    #   API server, in PEM format. A blank or nil value means that no custom CA certificates should be used.
+    # @option opts [String] :metrics_userid The name of the metrics database user.
+    # @option opts [String] :metrics_password The password of the metrics database user.
+    # @options opts [String] :metrics_server The host name or IP address of the metrics database server.
+    # @options opts [Integer] :metrics_port ('5432') The port number of the metrics database server.
+    # @options opts [String] :metrics_database The name of the metrics database.
+    # @return [Boolean] Returns `true` if the connection details and credentials are valid, or `false` otherwise.
+    #
+    def raw_connect(opts = {})
```
[the remark this answers is not in the document]
Comment (jhernand) [reply app/models/manageiq/providers/redhat/infra_manager/api_integration.rb]: No, this is the name that the UI uses, it has to be `raw_connect`.

File: app/models/manageiq/providers/redhat/infra_manager/api_integration.rb [hunk 3]
```diff
@@ -232,17 +219,132 @@ def process_api_features_support
...
+      check_connect_api(opts)
+      check_connect_metrics(opts)
```
[the remark this answers is not in the document]
Comment (jhernand) [reply app/models/manageiq/providers/redhat/infra_manager/api_integration.rb]: We need to check all the details that the user gives in the UI, if I understand correctly. We won't know what failed, nor do we need: the human user will know, hopefully, using her eyes and brain.

**Comment by borod108** [comment]
Well yes, we can safely assume the user has eyes and brain :) So using his
eyes and brain, will he be able to see which if the credentials fail? I
mean we have it in two tabs, and the validation button from the code here
looks like it sends the info from both, does it show in which tab the
failure is?
Its a UI question, I know, I am just wondering because I seem to remember
it did not work so well.
Any way I am for merging, just wondering...

**Comment by jhernand** [comment]
The tests are failing with this error:

```
     NoMethodError:
       undefined method `manager_ref' for #<ManageIQ::Providers::Redhat::InfraManager::Vm:0x0000000b45efd8>
       Did you mean?  manager_type
     # /home/travis/.rvm/gems/ruby-2.3.1/gems/activemodel-5.0.6/lib/active_model/attribute_methods.rb:433:in `method_missing'
     # ./spec/manageiq/app/models/manager_refresh/target_collection.rb:41:in `collect'
     # ./spec/manageiq/app/models/manager_refresh/target_collection.rb:41:in `id'
     # ./spec/manageiq/app/models/ems_refresh/refreshers/ems_refresher_mixin.rb:23:in `block (2 levels) in refresh'
     # ./spec/manageiq/app/models/ems_refresh/refreshers/ems_refresher_mixin.rb:23:in `each'
     # ./spec/manageiq/app/models/ems_refresh/refreshers/ems_refresher_mixin.rb:23:in `block in refresh'
     # ./spec/manageiq/app/models/ems_refresh/refreshers/ems_refresher_mixin.rb:14:in `each'
     # ./spec/manageiq/app/models/ems_refresh/refreshers/ems_refresher_mixin.rb:14:in `refresh'
     # ./spec/manageiq/app/models/manageiq/providers/base_manager/refresher.rb:9:in `refresh'
     # ./spec/manageiq/app/models/ems_refresh.rb:97:in `block in refresh'
     # ./spec/manageiq/app/models/ems_refresh.rb:96:in `each'
     # ./spec/manageiq/app/models/ems_refresh.rb:96:in `refresh'
     # ./spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_target_vm_4_spec.rb:147:in `block (2 levels) in <top (required)>'
```

I think that isn't related to this change, but rather to ManageIQ/manageiq#16405.

@Ladas @agrare @pkliczewski can you take a look. I think that the reason is that the above pull request assumes that all targeted objects have a `manager_ref` attribute, but virtual machines don't have it, at least not oVirt virtual machines.

**Comment by jhernand** [comment]
The fix for the issue that caused the test failure (ManageIQ/manageiq#16411) has been merged. I am closing and re-opening this to re-run the tests.

**Comment by agrare** [comment]
@simaishi don't backport this until https://github.com/ManageIQ/manageiq-providers-ovirt/pull/140 is also merged

**Comment by simaishi** [comment]
Gaprindashvili backport details:

```
$ git log -1
commit 0e4e22fea0e8d737ed6a8184137ee0ebf64cd84d
Author: Moti Asayag <masayag@redhat.com>
Date:   Wed Nov 8 09:09:21 2017 +0200

    Merge pull request #134 from jhernand/verify_metrics_details_from_raw_connect
    
    Check metrics details from `raw_connect`
    (cherry picked from commit af9c776d98752f8274d08e21b36163bff555e556)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1511079
```

**Pull request merged**